compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,265 @@
+    return rewriter.create<affine::AffineDelinearizeIndexOp>(loc, index, shape)
+        .getMultiIndex();
+  };
+
+  // For loop body:
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(forOp.getBody());
+    auto inductionVar = forOp.getInductionVar();
+    auto linearizedGatherIndices =
+        getGlobalGatherIndex(subgroupId, laneId, inductionVar);
+    auto delinearizedGlobalIndices =
+        delinearizeIndex(linearizedGatherIndices, sourceType.getShape());
+    auto linearizedBaseIndices =
+        getSubgroupStoreBaseIndex(subgroupId, inductionVar);
+    auto delinearizedLocalIndices =
+        delinearizeIndex(linearizedBaseIndices, localType.getShape());
+    rewriter.create<IREE::GPU::GlobalLoadDMAOp>(
+        loc, copy.getOperand(0), delinearizedGlobalIndices,
+        copy.getOutputs()[0], delinearizedLocalIndices);
+  }
+
+  // sync at the end of the loop across threads
+  rewriter.replaceOpWithNewOp<gpu::BarrierOp>(copy);
+  return success();
+}
+
+} // namespace
+
+static LogicalResult isEligibleForGlobalDMA(linalg::CopyOp copy) {
+  // source must be global address and target must be workgroup address.
+  auto sourceType = cast<MemRefType>(copy.getOperand(0).getType());
+  auto targetType = cast<MemRefType>(copy.getOutputs().front().getType());
+  if (!hasGlobalMemoryAddressSpace(sourceType)) {
+    LDBG("-- Op: "
+         << *copy
+         << "\n-- has source memory address space other than global.\n");
+    return failure();
+  }
+  if (targetType.getMemorySpace() !=
+      gpu::AddressSpaceAttr::get(copy->getContext(),
+                                 gpu::GPUDialect::getWorkgroupAddressSpace())) {
+    LDBG("-- Op: "
+         << *copy
+         << "\n-- has target memory address space other than workgroup.\n");
+    return failure();
+  }
+  auto useDMAConfig = getLoweringConfig<IREE::GPU::UseGlobalLoadDMAAttr>(copy);
+  if (!useDMAConfig) {
+    LDBG("-- Op: " << *copy
+                   << "\n-- does not have use_global_load_dma attribute.\n");
+    return failure();
+  }
+
+  // TODO: check that the copy's target memref is not a subview.
+  return success();
+}
+
+struct LowerToDMAPattern : public OpRewritePattern<linalg::CopyOp> {
+  LowerToDMAPattern(MLIRContext *context, ArrayRef<int64_t> workgroupSize,
+                    ArrayRef<int64_t> subgroupSize)
+      : OpRewritePattern<linalg::CopyOp>(context), workgroupSize(workgroupSize),
+        subgroupSize(subgroupSize) {}
+
+  LogicalResult matchAndRewrite(linalg::CopyOp copy,
+                                PatternRewriter &rewriter) const override {
+    if (failed(isEligibleForGlobalDMA(copy))) {
+      return failure();
+    }
+    auto result = distributeLinalgCopyToThreads(rewriter, copy, workgroupSize,
+                                                subgroupSize);
+    if (succeeded(result)) {
+      return success();
+    } else {
+      return failure();
+    }
```

Comment:
Should this be something like `return distributeLinalgCopyToThreads(...)`?

(It might not notify the rewriters, so the error message may not show up in `-debug`. This is something that I'm not familiar with, so I don't have a suggestion.)